inference-engine/src/mkldnn_plugin/nodes/non_max_suppression.cpp

```diff
@@ -135,110 +213,202 @@ class NonMaxSuppressionImpl: public ExtLayerBase {
-                    parallel_sort(scores_vector.begin(), scores_vector.end(),
-                        [](const std::pair<float, int>& l, const std::pair<float, int>& r) { return l.first > r.first; });
-
-                    int io_selection_size = 1;
-                    fb.push_back({ scores_vector[0].first, batch, class_idx, scores_vector[0].second });
-                    for (int box_idx = 1; (box_idx < static_cast<int>(scores_vector.size()) && io_selection_size < max_output_boxes_per_class); box_idx++) {
-                        bool box_is_selected = true;
-                        for (int idx = io_selection_size - 1; idx >= 0; idx--) {
-                            float iou = intersectionOverUnion(&boxesPtr[scores_vector[box_idx].second * 4],
-                                             &boxesPtr[scores_vector[idx].second * 4], center_point_box);
-                            if (iou > iou_threshold) {
-                                box_is_selected = false;
-                                break;
+                                if (iou >= iou_threshold) {
+                                    box_is_selected = false;
+                                    break;
+                                }
+
+                                if (currBox.score <= score_threshold)
+                                    break;
                             }
-                        }
 
-                        if (box_is_selected) {
-                            scores_vector[io_selection_size] = scores_vector[box_idx];
-                            io_selection_size++;
-                            fb.push_back({ scores_vector[box_idx].first, batch, class_idx, scores_vector[box_idx].second });
+                            currBox.suppress_begin_index = fb.size();
+
+                            if (box_is_selected) {
+                                if (currBox.score == origScore) {
+                                    fb.push_back({ currBox.score, batch, class_idx, currBox.idx });
+                                    continue;
+                                }
+                                if (currBox.score > score_threshold) {
+                                    sorted_boxes.push(currBox);
+                                }
+                            }
                         }
                     }
+                    numFiltBox[batch][class_idx] = fb.size();
+                    size_t offset = batch*num_classes*max_output_boxes_per_class + class_idx*max_output_boxes_per_class;
+                    cpu_memcpy(filtBoxes.data() + offset, fb.data(), fb.size() * sizeof(filteredBoxes));
                 }
             }
+        });
+
+        size_t startOffset = numFiltBox[0][0];
+        for (size_t b = 0; b < numFiltBox.size(); b++) {
+            for (size_t c = (b == 0 ? 1 : 0); c < numFiltBox[b].size(); c++) {
+                size_t offset = b*num_classes*max_output_boxes_per_class + c*max_output_boxes_per_class;
```

Comment:
b*num_classes*max_output_boxes_per_class may be calculated outside current loop.

Reply:
fixed